robot tests improvement

jburel opened the pull request:
This PR uses
 * ``robotframework-selenium2library==3.0.0``
* ``robotframework==3.0.2``

This PR:
* removes usage of deprecated keywords
* removes usage of reserve syntax
* splits the tests in smaller tests
* splits some test suite i.e. file into smaller suite
* adds new tasks to allow rerun failing tests only or failing suites
* adds new tasks to merge results
* adds new tasks to run the script in parallel using ``robotframework-pabot``
* re-activates commented out tests
* review search tests. Test not testing UI functionality will be converted in an integration test in another PR

Note that the number of tests has increased. The number in the job
https://ci.openmicroscopy.org/job/OMERO-DEV-merge-robotframework/ was the aggregate of 2 runs.


We could split a bit more the tests but that could be done in follow up PRs
or improve passing one e.g. removing usage of Sleep (not great way to synch)
The ``robotframework-pabot`` is still in its early development

Devspace running at https://10.0.51.135:32784/

# Related PRs
 * https://github.com/openmicroscopy/ansible-role-devspace/pull/7
 * https://github.com/openmicroscopy/devspace/pull/88

# Related reading

https://trello.com/c/QgTG93Ow/1-devspace

cc @will-moore @rgozim @bramalingam

rgozim committed:
Added wait for page load in robot test login

will-moore committed:
Fix robot 'id=choose_file_anns' not visible

will-moore committed:
Remove 'Maximize Browser Window' from Suite Setup

jburel committed:
Attempt to fix teardown failure

jburel committed:
Attempt to fix tear down issue

jburel committed:
Remove Get Element attribute warning

jburel committed:
Remove remaining warning

jburel committed:
Remove usage of future reserve syntax

will-moore committed:
Wait for Save of 'Owner Group Edit' robot test

will-moore committed:
Improve reliability of robot 'Test Search Results File Annotations'

jburel committed:
Merge pull request #14 from will-moore/robot-infra

Robot infra

jburel committed:
Split rating test

jburel committed:
Attempt to stabilize rating filter test

jburel committed:
Attempt to stabilize tests

jburel committed:
Update annotation test

jburel committed:
Re-organize tests

jburel committed:
Already tested in tagging_test

jburel committed:
Fix typo

jburel committed:
change locator name

jburel committed:
Rename image

jburel committed:
Clean up some tests

jburel committed:
Fix setup

jburel committed:
Split tests

jburel committed:
Remove check

jburel committed:
Test search

jburel committed:
Attempt to stabilize rating tests

jburel committed:
Clean up search test

The keywords check is moved to an integration test

jburel committed:
Fix filter by rating

jburel committed:
Split tests and fix broken ones

Double click is not working on FF

jburel committed:
Fix tagging tests

jburel committed:
Add option to run tests using pabot

jburel committed:
Delete projects and annotations after tests

jburel committed:
Update keywords

jburel committed:
Add requirement

jburel committed:
increase timeout to stabilize on chrome

jburel committed:
Add merge-results task

jburel committed:
Bump number of processors

jburel committed:
Add new targets

jburel committed:
add rerun with pabot target

jburel committed:
Add documentation

jburel committed:
Split the test suite

jburel committed:
Rename file, split suite and re-activate tests

jburel committed:
Clean up

jburel committed:
Remove usage of deprecated keyword

jburel committed:
Remove commented out test

This is tested in view_image.txt

jburel committed:
Testing rerunfailedsuites

jburel committed:
Simplify test

Loading tested in other tests

jburel committed:
Add option to run the failed suites or the failed tests

jburel committed:
Re-organise tests and add sleep to stabilize test

jburel committed:
Split the test suites and re-organize tests

jburel committed:
Update doc

jburel committed:
Add new keyword

jburel committed:
Re-organize tests

jburel committed:
Review tests

jburel committed:
Slow down the test

jburel committed:
Fix variables

jburel committed:
Update doc

jburel committed:
Split test

jburel committed:
Clean up tests

jburel committed:
Move check

jburel committed:
Combine tasks to over more options

jburel committed:
Fix delete rating test

jburel committed:
Add info

jburel committed:
Attempt to stabilize delete rating

will-moore commented:
Does the README need updating? E.g. we no-longer want to do
```
 ./build.py -f components/tests/ui/build.xml web-firefox -DTEST=tree_test.txt
```

will-moore commented on components/tests/ui/build.xml:
```diff
@@ -237,7 +308,8 @@
     </target>
 
-    <!-- test with chrome This requires the chrome driver to be installed-->
+    <!-- test with chrome This requires the chrome driver to be installed -->
```
... to *be* installed

will-moore commented on components/tests/ui/build.xml:
```diff
@@ -264,7 +336,176 @@
+    </target>
+
+    <!-- Rerun the tests that failed using pabot. It is possible to retest in another broswer
+    either the test failures or the full suite. If chrome is used, the chrome driver has be installed -->
```
Again ...has *to* be installed

will-moore commented on components/tests/ui/testcases/web/search_annotation_test.txt:
```diff
@@ -0,0 +1,50 @@
+
+    # Annotate single Screen
+    # NB: Attachments panel may be open from previous test. Try opening 'closed' panel in case...
+    Run Keyword And Ignore Error                Click Element  xpath=//h1[@data-name='attachments'][contains(@class, 'closed')]
```
Can probably replace this line and the comment line above now, since this we won't have variable state from previous test.
Instead, we will always need to expand the annotations panel with
```
Click Element                               xpath=//h1[@data-name='attachments']
```

will-moore commented:
Looks good. Didn't read every test, but the re-organisation / splitting of tests looks good.

Would be nice to have more in the README about the newer build test options, particularly
those that it would be useful to run locally.

jburel committed:
Describe new tasks

jburel committed:
Review documentation and simplify deprecated target

jburel committed:
Mark the cli/insight tasks as deprecated

jburel committed:
Minor edit

jburel committed:
Adjust test

jburel commented:
@will-moore I have updated the readme as suggested
and updated the test

I have also indicated that the targets related to insight and cli are deprecated

will-moore commented on components/tests/ui/README.md:
````diff
@@ -225,10 +225,10 @@ components/tests/ui/target/reports/web/firefox/
 
 ```
-./build.py -f components/tests/ui/build.xml web-firefox -DTEST=tree_test.txt
+./build.py -f components/tests/ui/build.xml web-browser -DTEST=tree_test.txt
````
this file doesn't exist (hasn't for a while I think)

will-moore commented on components/tests/ui/README.md:
````diff
@@ -237,6 +237,49 @@ with the ouput being available under
 ```
 
+It is possible to rerun tests that failed either in a different browser or the same browser.
+By default, the tests will be first run in Firefox and rerun in chrome.
````
Chrome (upper case)

will-moore commented on components/tests/ui/README.md:
```diff
@@ -237,6 +237,49 @@ with the ouput being available under
+
+``
+
+To merge the outputs after running a `rerun` task first on Firefox  use
```
Two spaces in ```Firefox  use```

will-moore commented:
I have these installed:
```
$ pip freeze | grep robot
robotframework==3.0.2
robotframework-extendedselenium2library==0.9.1
robotframework-selenium2library==3.0.0
robotframework-seleniumlibrary==3.0.1
```
The README only says you need ```robotframework-selenium2library```
but if I  ```pip install robotframework-seleniumlibrary``` then tests fail with
``` No keyword with name 'Open Browser' found. ```

jburel commented:
* robotframework-extendedselenium2library==0.9.1 that's the one you have installed when trying to fix an issue but it did not support v 3.0.0 so you can remove it

You should only need to install
``
robotframework==3.0.2
robotframework-selenium2library==3.0.0
``

``
robotframework-seleniumlibrary==3.0.1
`` 
is installed when installing ``robotframework-selenium2library``

jburel committed:
Fix typo and update example

will-moore commented:
Ah, it was the ```extendedselenium2library``` that was causing problems. Uninstalled it. Looking better....

will-moore committed:
Improve robot ratings tests

jburel committed:
Merge pull request #15 from will-moore/robot-infra

Improve robot ratings tests

rgozim committed:
Updated readme

jburel committed:
Merge pull request #16 from rgozim/robot-infra

Updated readme

jburel commented:
Having the robot tests green, indicates that this PR https://github.com/openmicroscopy/openmicroscopy/pull/5631 broke the tests this morning

https://10.0.51.135:32784/job/OMERO-robot/5/

jburel commented:
Any more comments @will-moore @rgozim?
We now have an example PR "breaking the tests" :-)

jburel commented:
Discussed with @will-moore and @rgozim merging this PR
Further improvements will happen in follow up PRs
